packages/core/src/sync-realtime/service.ts

```diff
@@ -79,19 +94,48 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
     this.requestQueue = requestQueue;
     this.sources = sources;
 
-    this.queue = this.buildQueue();
+    this.hasFactorySource = sources.some(sourceIsFactory);
+    this.logFilterSources = sources.filter(sourceIsLogFilter);
+    this.factorySources = sources.filter(sourceIsFactory);
+
+    const isAddressDefined = this.logFilterSources.every(
+      (source) => !!source.criteria.address,
+    );
+    this.address =
+      !this.hasFactorySource && isAddressDefined
+        ? (sources.flatMap((source) => source.criteria.address) as Address[])
+        : undefined;
+
+    this.topics = sources.flatMap((source) => {
+      const topics: Hex[] = [];
+
+      if (sourceIsFactory(source)) {
+        topics.push(source.criteria.eventSelector);
+      }
+
+      const topic0 = source.criteria.topics?.[0];
+      if (topic0) {
+        if (Array.isArray(topic0)) topics.push(...topic0);
+        else topics.push(topic0);
+      } else {
+        topics.push(...(Object.keys(source.abiEvents.bySelector) as Hex[]));
+      }
+      return topics;
+    });
   }
 
   setup = async () => {
     // Initialize state variables. Required when restarting the service.
     this.blocks = [];
 
     // Fetch the latest block, and remote chain Id for the network.
-    let latestBlock: BlockWithTransactions;
+    let latestBlock: RealtimeBlock;
+    let finalizedBlock: RealtimeBlock;
     let rpcChainId: number;
     try {
-      [latestBlock, rpcChainId] = await Promise.all([
-        this.getLatestBlock(),
+      [latestBlock, finalizedBlock, rpcChainId] = await Promise.all([
+        this._eth_getBlockByNumber("latest"),
+        this._eth_getBlockByNumber("finalized"),
```

Comment:
Not all networks support this, and I suspect the behavior might be inconsistent across providers (some might error, others might return `"latest"` instead, etc.

We could still consider using it if we think there's a real benefit, but would need to handle these cases and fall back to the old behavior.